### PR TITLE
feat: self-service API keys + RBAC simplification (admin/user) + per-key rate limit

### PR DESCRIPTION
**Pull request by JinY0ung-Shin**
## Summary

세 가지 상호 연관된 변경을 한 묶음으로 구현합니다 (스펙: `docs/superpowers/specs/2026-05-29-self-service-apikey-and-rbac-simplification-design.md`).

### A. RBAC 단순화 (admin / user)
- 역할을 `admin`(전체 권한) / `user`(`gateway.monitoring.read`, `alerts.read`, `apikeys.self`) 2개로 축소. `developer`/`viewer` 제거.
- `ALL_PERMISSIONS`에 `apikeys.self` 추가, `ROLE_PRIORITY = ["admin", "user"]`, `CurrentUser.sub`(Keycloak sub) 추가.
- `_seed_roles`가 admin/user를 시드하고 스테일 시스템 역할을 prune.
- `keycloak/realm-export.json` realm roles 교체 + 기존 developer/viewer 사용자 → user 다운그레이드 스크립트.

### B. 자가발급 API Key (self-service)
- JWT 사용자가 본인 키 1개 직접 관리: `GET/POST/DELETE /admin/api-keys/me` + `POST /me/regenerate` (`apikeys.self`, JWT 전용, `owner == sub`).
- 고정 스코프: `allowed_databases=['*']`, `allowed_routes=['query-api','s3-api']`(llm 제외), 30 req/min. consumer `self_{sub}`, admin 키 `self_` 접두사 예약.
- 동시 생성 race는 DB 행을 먼저 claim(unique consumer_name)해 직렬화, APISIX 프로비저닝 실패 시 롤백. "My API Key" UI.

### C. admin 발급 키 분당 한도 조절
- `ApiKeyAccess`에 `owner` + `rate_limit_per_minute` 컬럼(Alembic 0007). APISIX `limit-count` 키별 동적 프로비저닝. admin UI 입력 필드.
- 업데이트는 `model_fields_set`로 생략 vs 명시적 null 구분 — 명시적 null은 한도 해제(limit-count 제거 + DB NULL). 기존 consumer plugin을 읽지 못하면 lossy PUT 대신 502.

## Test Plan
- [x] 백엔드 전체 **1280 passed** / 프론트 **335 passed** / 프론트 빌드 통과
- [x] Alembic 0007 round-trip 검증
- [x] 자가발급 키 생성/재생성/삭제, 1인1키(409), 동시 race(409 no-clobber), APISIX 실패 롤백, JWT 가드, `['*']` 와일드카드
- [x] 키별 limit-count(create/update, key-auth 보존, 명시적 null 해제, 읽기 실패 502)
- [ ] (배포 시) 기존 Keycloak developer/viewer 사용자에 `scripts/migrate_roles_to_user.py` 실행

## Known limitations
- 라우트 deny-by-default 하드닝 보류(현재 전체 허용 방침). llm-proxy/llm-admin은 키 동기화로 자가발급 키가 명시 차단됨.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: unibridge-service/app/routers/api_keys.py [hunk 1]
```diff
@@ -279,6 +461,8 @@ async def update_api_key(
         access.allowed_databases = json.dumps(body.allowed_databases) if body.allowed_databases else None
     if body.allowed_routes is not None:
         access.allowed_routes = json.dumps(body.allowed_routes) if body.allowed_routes else None
+    if body.rate_limit_per_minute is not None:
```
Comment (JinY0ung-Shin) [inline unibridge-service/app/routers/api_keys.py]: This makes the documented/admin-UI "empty = unlimited" flow ineffective for updates. The UI sends `rate_limit_per_minute: null`, but this guard skips both the DB update and APISIX plugin removal, so an existing limit remains active while the form implies it was cleared. Please distinguish omitted from explicitly-null, for example by checking whether the field was provided and then removing `limit-count`/storing `NULL` when it is `None`.

File: unibridge-service/app/routers/api_keys.py [hunk 2]
```diff
@@ -224,6 +271,133 @@ async def create_api_key(
...
+        description=f"Self-service key for {user.username}",
+        owner=user.sub,
+        allowed_databases=json.dumps(SELF_ALLOWED_DATABASES),
+        allowed_routes=json.dumps(SELF_ALLOWED_ROUTES),
+        rate_limit_per_minute=SELF_RATE_LIMIT,
+    )
+    db.add(access)
+    try:
+        await db.commit()
+    except IntegrityError:
```
Comment (JinY0ung-Shin) [inline unibridge-service/app/routers/api_keys.py:322]: If this commit fails after the APISIX consumer was created/updated, the external state is not rolled back. In a concurrent self-key create, the second request can overwrite the same `self_<sub>` consumer with a new key, then hit this `IntegrityError` and return 409, leaving the first successful DB row pointing at a consumer whose key no longer matches the key returned to the user. Please clean up or restore the APISIX consumer and route restrictions on DB commit failure, and add a regression test for this path.

**Review by JinY0ung-Shin — COMMENTED** [review]
Two issues need fixes before this is safe to merge. The CI is green, but these are contract/concurrency gaps not currently covered by tests.

File: unibridge-service/app/routers/api_keys.py (hunk 1 above)
Comment (JinY0ung-Shin) [reply unibridge-service/app/routers/api_keys.py]: Fixed in 9109dd5. `update_api_key` now distinguishes omitted from explicit-null via `"rate_limit_per_minute" in body.model_fields_set` instead of `is not None`. When the field is explicitly provided as null: `_build_limit_count_plugin(None)` returns None → the APISIX `limit-count` plugin is popped, and the DB row is set to NULL. Omitting the field preserves the existing limit (no consumer PUT, no DB change). Other ApiKeyUpdate fields keep their `is not None` = no-change semantics. Tests added: test_update_clears_rate_limit_when_explicit_null (asserts limit-count popped + DB null) and test_update_omitted_rate_limit_preserves_existing.

File: unibridge-service/app/routers/api_keys.py (hunk 2 above)
Comment (JinY0ung-Shin) [reply unibridge-service/app/routers/api_keys.py:322]: Fixed in 9109dd5. `create_my_api_key` now claims the DB row BEFORE any APISIX mutation: it inserts+commits the ApiKeyAccess row first, and the unique `consumer_name` constraint (deterministic `self_<sub>`) serializes concurrent creates — the loser fails with IntegrityError → 409 before `put_resource` is ever called, so it can no longer clobber the winner's consumer key. APISIX provisioning (consumer + consumer-restriction) runs only after the commit succeeds; on provisioning failure the DB row is rolled back (delete+commit) and the consumer is best-effort deleted, then the error is re-raised. Regression tests added: test_post_me_integrity_error_returns_409_without_apisix_clobber (asserts 409 AND put_resource was never called) and test_post_me_rolls_back_db_on_apisix_failure (asserts 5xx, GET /me → null, delete_resource attempted). One residual: a short self-correcting window exists between the DB commit and APISIX provisioning where GET /me returns the row with a null masked key — the GET handler already tolerates this.

File: unibridge-service/app/routers/api_keys.py [hunk 3]
```diff
@@ -241,20 +424,33 @@ async def update_api_key(
...
+    # omitted field means "no change". model_fields_set holds the field names
+    # that were actually present in the request payload.
+    rate_limit_provided = "rate_limit_per_minute" in body.model_fields_set
+    if body.api_key or rate_limit_provided:
         try:
             existing_consumer = await apisix_client.get_resource("consumers", name)
-            existing_plugins = existing_consumer.get("plugins", {})
+            existing_plugins = dict(existing_consumer.get("plugins", {}))
             old_consumer_plugins = dict(existing_plugins)  # snapshot for rollback
         except Exception:
```
Comment (JinY0ung-Shin) [inline unibridge-service/app/routers/api_keys.py]: This fallback still looks unsafe for update payloads that need to preserve existing consumer plugins. If `get_resource` fails transiently, a rate-limit-only update continues with `{}` and PUTs a consumer with only `limit-count` (or no plugins for explicit-null), dropping the existing `key-auth`. Similarly, an api-key-only rotation can silently drop an existing `limit-count`. Please fail the request when the existing consumer plugins cannot be fetched for a preservation-sensitive update, or rebuild the full plugin set from a reliable source before PUT. A regression test with `get_resource` raising during a rate-limit-only update would cover this.

**Review by JinY0ung-Shin — COMMENTED** [review]
Rechecked the two earlier threads on 9109dd5: both original issues are addressed and the targeted regression tests pass locally. I found one remaining edge case in the same update path that should be fixed before merge. Also, the PR body still lists "admin key updates cannot clear rate limit" as a known limitation, which is now stale after this fix.

File: unibridge-service/app/routers/api_keys.py (hunk 3 above)
Comment (JinY0ung-Shin) [reply unibridge-service/app/routers/api_keys.py]: Fixed in 51a9ff7. The unsafe `except: existing_plugins = {}` fallback is removed. `update_api_key` now raises 502 when the existing consumer plugins can't be fetched, so a preservation-sensitive update (rate-limit-only or api-key-only) never PUTs a partial plugin set that would drop `key-auth` or `limit-count`. The raise happens before any `put_resource`, so no lossy write can occur. Tests: the old empty-plugins test was replaced by test_update_api_key_fails_when_existing_consumer_unreadable (asserts 502 + no consumer PUT), and a new test_update_rate_limit_only_fails_when_consumer_unreadable covers the rate-limit-only path with get_resource raising (asserts 502 + key-auth never dropped). Also updated the PR body to drop the now-stale 'cannot clear rate limit' limitation.
